src/util/kaldi-table-inl.h

```diff
@@ -245,52 +282,81 @@ template<class Holder>  class SequentialTableReaderScriptImpl:
     } else {
       return true;
     }
+    // Possible states                                          Return value
+    // kLoadedSucceeded/kRangeExtracted/kRangeExtractionFailed  true
+    // kError (if opts_.permissive)                             true
+    // kError (if !opts_.permissive)                            false
+    // kEof (if script_input_.Close() && !opts.permissive)      false     
+    // kEof (if !script_input_.Close() || opts.permissive)      true     
+    // kUninitialized/kFileStart/kHaveScpLine                   true
+    // kUnitialized                                             true
   }
 
   virtual ~SequentialTableReaderScriptImpl() {
-    if (state_ == kError)
-      KALDI_ERR << "Reading script file failed: from scp "
-                << PrintableRxfilename(script_rxfilename_);
-    // If you don't want this exception to be thrown you can
-    // call Close() and check the status.
-    if (state_ == kLoadSucceeded)
-      holder_.Clear();
+    if (this->IsOpen() && !Close())
+      KALDI_ERR << "TableReader: reading script file failed: from scp "
+                      << PrintableRxfilename(script_rxfilename_);
   }
  private:
-  bool LoadCurrent() {
+  bool EnsureObjectLoaded() {
     // Attempts to load object whose rxfilename is on the current scp line.
-    if (state_ != kHaveScpLine)
-      KALDI_ERR << "LoadCurrent() called at the wrong time.";
-    bool ans;
-    // note, NULL means it doesn't read the binary-mode header
-    if (Holder::IsReadInBinary()) {
-      ans = data_input_.Open(data_rxfilename_, NULL);
-    } else {
-      ans = data_input_.OpenTextMode(data_rxfilename_);
-    }
-    if (!ans) {
-      // May want to make this warning a VLOG at some point
-      KALDI_WARN << "Failed to open file "
-                 << PrintableRxfilename(data_rxfilename_);
-      state_ = kLoadFailed;
-      return false;
-    } else {
-      if (holder_.Read(data_input_.Stream())) {
-        state_ = kLoadSucceeded;
-        return true;
-      } else {  // holder_ will not contain data.
-        KALDI_WARN << "Failed to load object from "
+    if (state_ != kLoadSucceeded) {
```

Comment:
Now that we are calling EnsureObjectLoaded() from Value() without first checking the state, we have to be prepared for it this function to be called twice.  If the state is kRangeExtracted or kRangeExtractionFailed, this function should just return without doing anything, because everything that needs to be done is already done.
And I think this if-statement (which I know is old code) could be reorganized for more clarity to say
  if (state_ == kHaveScpLine) { 
    ... do all this stuff
  } else if (state != kLoadSucceeded) {
     KALDI_ERR << "EnsureObjectLoaded() called at the wrong time.";
  }